Scraper

odpaleniek1337 opened the pull request:
For now it cannot scan the offer_type from the main page, it also sometimes gets polish translations.

odpaleniek1337 committed:
initial webscraper

odpaleniek1337 committed:
scraper can now dynamically work with paginated results

odpaleniek1337 committed:
added scraping items from otd site; added dumping to csv; added checking if all fields are present; added scraping links

odpaleniek1337 committed:
added scraping items from another ref site; fixed artifact newlines; added loading/saving corrupted data from one site with the use of pickling

odpaleniek1337 committed:
Merge branch 'main' into scraper

odpaleniek1337 committed:
Update requirements.txt

krasnowsky commented on scraper/scraper_utils.py:
```diff
@@ -0,0 +1,116 @@
+
+        for item in filtered_data.keys():
+            if 'floor' in item:
+                filtered_data['floor'] = filtered_data['floor'].split(':')[1].strip('"}flor_')
```
misspell or is it really 'flor_' XD?

odpaleniek1337 replied:
it strips chars, so no need to type other o xD

danielbaraniak replied:
Floor higher than 11 have value `floor_higher_10`. I think we can just leave it without striping and take care of it in `ml.src.data`.

odpaleniek1337 replied:
I would say either we parse it on scraper side or pass everything unparsed in csv.
Since **artifact.csv** is a csv, some unscraped strings may contain comma and then corrupt file

krasnowsky commented on scraper/scraper_utils.py:
```diff
@@ -0,0 +1,116 @@
+            if SEARCHED_FIELDS_OTD['offer_type'] in item:
+                filtered_data['offer_type'] = item.split(':')[1].strip('"}')
+            if SEARCHED_FIELDS_OTD['floor'] in item:
+                filtered_data['floor'] = item.split(':')[1].strip('[]"_flor')
```
same as above

odpaleniek1337 replied:
same as above

krasnowsky commented on scraper/scraper_utils.py:
```diff
@@ -0,0 +1,116 @@
+            if SEARCHED_FIELDS_OTD['floor'] in item:
+                filtered_data['floor'] = item.split(':')[1].strip('[]"_flor')
+            if SEARCHED_FIELDS_OTD['area'] in item:
+                filtered_data['area'] = item.split(':')[2].strip('[]"_flor')
```
same as above

odpaleniek1337 replied:
same as above

krasnowsky commented on scraper/sprinters_driver.py:
```diff
@@ -0,0 +1,88 @@
+
+class SprintersDriver:
+    def __init__(self) -> None:
+        browser = 1#get_random_number_from_range(1, 3)
```
should browser always be 1?

odpaleniek1337 replied:
tested it on the chrome browser, downloads, launching worked as well on my machine, but to speed things up deside to choose 1, i know for now it works fine on browser one

odpaleniek1337 replied:
tried also browser 2 and 3 in limited environment and browser 3 works fine as well (probably because it also works on Chrome webdriver)

browser 2 raised error below while trying to close cookies using second path -> preferences -> save preferences

`Message: element click intercepted: Element is not clickable at point (285, 963)`

closing cookies using first path works tho

odpaleniek1337 replied:
done in https://github.com/230443/Sprinters/pull/15/commits/9e42bd68aef8258c4998e4a938065f396967f483

krasnowsky commented on scraper/sprinters_driver.py:
```diff
@@ -0,0 +1,88 @@
+
+    def close_cookies(self, cookies_config_routes: List[str]) -> bool:
+        self.driver.implicitly_wait(3)
+        which_route = 2#get_random_number_from_range(1, 2)
```
should it always be 2?

odpaleniek1337 replied:
fixed in [2411137debdaafa98096a6c164059f991d8c3b3c ](https://github.com/230443/Sprinters/pull/15/commits/2411137debdaafa98096a6c164059f991d8c3b3c) 
it works both ways

krasnowsky commented on scraper/scraper.py:
```diff
@@ -0,0 +1,65 @@
+
+def main():
+    configparser = ConfigParser()
+    configparser.read('config.ini')
```
what do u think about moving this config to main config file? (settings.cfg in the main directory)
@odpaleniek1337 @230443

odpaleniek1337 replied:
it depends for me its cool, but we need to think if all the components would be packed onto one machine & directory
if so, we could do that, otherwise i think its good if all components have their own configs

danielbaraniak replied:
I would prefer it to be consistent for all modules. So:
1. Create separate configs for all modules
2. One global config.

I think the second option is good enough even if we run it separately since we can pull the repository and configure only required part on scraping machine. And its easier to manage when using on one machine.

However, I think its good that we have separate `requirements.txt` and it would be good if ml and web also had it separated, so we do not install unnecessary stuff.

odpaleniek1337 replied:
requirements are updated in  **scraper** folder

okay, im moving **scraper** config onto the main directory config

odpaleniek1337 replied:
done in https://github.com/230443/Sprinters/pull/15/commits/05dbe6ae6ebf1a0a9a46985df1d9f065dfbf703a

krasnowsky commented on scraper/scraper.py:
```diff
@@ -0,0 +1,65 @@
+    dump_csv_dict_data(flats_data)
+
+if __name__ == '__main__':
+    main()
```
Maybe we can do the same thing as we did with ml and backend, move running it into launch.py, for example with parameter --scrap

    `parser.add_argument(
        '--train', action='store_true', help='Create machine learning model.'
    )
    parser.add_argument('--web', action='store_true', help='Start web server.')
    parser.add_argument('--scrap', action='store_true', help='Scrap data form websites.')

    args = parser.parse_args()

    if args.train:
        orchestrator.main()
    elif args.web:
        from webservice.app import app
        app.run(debug=True)
     elif args.scrap:
        run scrapper here

    else:
        parser.print_help()`

odpaleniek1337 replied:
up, if we decide we run all things on one machine, allright

danielbaraniak replied:
I think we can add the option which invokes `scraper.scraper.main` in launch.py. It won't affect modularity much.

odpaleniek1337 replied:
done in https://github.com/230443/Sprinters/pull/15/commits/05dbe6ae6ebf1a0a9a46985df1d9f065dfbf703a

odpaleniek1337 committed:
Update sprinters_driver.py

scraper can now randomly choose whether to accept cookies or prolong it by clicking see preferences

danielbaraniak commented on scraper/scraper_utils.py:
```diff
@@ -0,0 +1,116 @@
+
+    return check_if_data_contains_all_fields(filtered_data, SEARCHED_FIELDS_OX)
+
+def get_flat_otd_data(link: str) -> dict:
+    request = requests.get(link)
+    str_request_data = str(request.content).split(',')
+    filtered_data = {}
+    try:
+        for item in str_request_data:
+            if SEARCHED_FIELDS_OTD['offer_type'] in item:
+                filtered_data['offer_type'] = item.split(':')[1].strip('"}')
+            if SEARCHED_FIELDS_OTD['floor'] in item:
+                filtered_data['floor'] = item.split(':')[1].strip('[]"_flor')
+            if SEARCHED_FIELDS_OTD['area'] in item:
+                filtered_data['area'] = item.split(':')[2].strip('[]"_flor')
+            if SEARCHED_FIELDS_OTD['rooms'] in item:
+                filtered_data['rooms'] = item.split(':')[1].strip('[]"')
+            if SEARCHED_FIELDS_OTD['offer_type_of_building'] in item:
+                filtered_data['offer_type_of_building'] = item.split(':')[1].strip('[]"')
+            if SEARCHED_FIELDS_OTD['market'] in item:
+                filtered_data['market'] = item.split(':')[1].strip('[]"')
+            if SEARCHED_FIELDS_OTD['longtitude'] in item:
+                filtered_data['longtitude'] = item.split(':')[1].strip('[]"')
+            if SEARCHED_FIELDS_OTD['latitude'] in item:
+                filtered_data['latitude'] = item.split(':')[1].strip('[]"')
+            if SEARCHED_FIELDS_OTD['price'] in item:
+                filtered_data['price'] = item.split(':')[1].strip('[]"')
+    except Exception:
+        dump_scraped_data_to_pickle(str_request_data, str(uuid.uuid4()))
+    
+    return check_if_data_contains_all_fields(filtered_data, SEARCHED_FIELDS_OTD)
```
Those `if`s are quite unreadable. It should be possible to parse json and access elements of interest by key, which would simplify code a little.

```python
text = response.content.decode('utf-8')
m = re.search(r'<script id="__NEXT_DATA__".*>(.*)</script>', text)
data = json.loads(m.groups()[0])
target = data['props']['pageProps']['ad']['target'] | data['props']['pageProps']['adTrackingData']

SEARCHED_FIELDS_OTD_TRANSLATION = {
    'price': 'ad_price',
    "offer_type": 'user_type',
    "floor": 'Floor_no',
    "area": 'Area',
    "rooms": 'Rooms_num',
    "offer_type_of_building": 'ProperType',
    "market": 'market',
    "longtitude": 'long',
    "latitude": 'lat',
    'city_name': 'city_name',
    'build_year': 'Build_year'
}

result = {}
for target_key, raw_key  in SEARCHED_FIELDS_OTD_TRANSLATION.items():
    v = target.get(raw_key)
    if isinstance(v, list):
        v = v[0]
    result[target_key] = v
```

or just:

```python
result2 = {
    'price': data['props']['pageProps']['adTrackingData']['ad_price'],
...
}
```

danielbaraniak commented on scraper/.gitignore:
```diff
@@ -0,0 +1 @@
+config.ini
```
nit: no newline at the end of the file.

odpaleniek1337 replied:
fixed in https://github.com/230443/Sprinters/pull/15/commits/832085a3c3d729274b6b57a6662e134c1027af6a

danielbaraniak commented on scraper/scraper_utils.py:
```diff
@@ -0,0 +1,116 @@
+def loadscraped_data_from_pickle(filename: str):
+    with open(filename, 'rb') as input_file:
+        loaded_data = pickle.load(input_file)
+    return loaded_data
```
nit: no newline at the end of the file.

odpaleniek1337 replied:
fixed in https://github.com/230443/Sprinters/pull/15/commits/832085a3c3d729274b6b57a6662e134c1027af6a

odpaleniek1337 committed:
fixed ending newlines

odpaleniek1337 committed:
moved scraping config into main config file; added scraper config to template; added scraper to main launch.py with '--scrap' argument; standarised launch.py; added 'usage' scraper part into main readme; standarised readme;fixed and standarised relative imports in scraper; disabled ability to run scraper standalone - imports, config loading doesnt work

odpaleniek1337 committed:
switched embedded implicit waits to regular time sleep since driver waits dont work as intented

230443 commented on README.md:
````diff
@@ -39,4 +39,12 @@ Competance project
+
+- Start scraping worker
+    ```bash
+    python3 -m launch.py --scraper
````
```suggestion
    python3 -m launch.py --scrape
```

danielbaraniak replied:
```suggestion
    python3 -m launch --scrape
```

danielbaraniak commented on launch.py:
```diff
@@ -2,23 +2,26 @@
 
+    parser.add_argument('--train', action='store_true', help='Create machine learning model.')
+    parser.add_argument('--web', action='store_true', help='Start web server.')
+    parser.add_argument('--scrap', action='store_true', help='Invoke scraping worker.')
```
```suggestion
    parser.add_argument('--scrape', action='store_true', help='Invoke scraping worker.')
```

danielbaraniak reviewed (changes requested):
different spelling in readme and launch.py

odpaleniek1337 committed:
great regexes

odpaleniek1337 committed:
added requestes changes to readme

odpaleniek1337 committed:
fixes in launch.py; fix in scraper

odpaleniek1337 committed:
changed default location of artifact to ml/data/raw
